sw/device/bsp/w25q/w25q.c

```diff
@@ -0,0 +1,1247 @@
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Define the status variable
+    w25q_error_codes_t status;
+
+    if (length < RX_DMA_THRESHOLD) {
+        status = w25q128jw_read_quad(addr, data, length);
+        if (status != FLASH_OK) return status;
+    } else {
+        // Wait DMA to be free
+        while(!dma_is_ready());
+        status = w25q128jw_read_quad_dma(addr, data, length);
+        if (status != FLASH_OK) return status;
+    }
+
+    return FLASH_OK;
+}
+
+uint8_t w25q128jw_write(uint32_t addr, void *data, uint32_t length, uint8_t erase_before_write) {
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Define the status variable
+    w25q_error_codes_t status;
+
+    if (erase_before_write == 1) {
+        status = erase_and_write(addr, data, length);
+        if (status != FLASH_OK) return status;
+    } else {
+        // Wait DMA to be free
+        while(!dma_is_ready());
+        status = w25q128jw_write_quad_dma(addr, data, length);
+        if (status != FLASH_OK) return status;
+    }
+
+    return FLASH_OK;
+}
+
+uint8_t w25q128jw_read_standard(uint32_t addr, void* data, uint32_t length) {
+    // Sanity checks
+    if (sanity_checks(addr, data, length) == 0) return FLASH_ERROR;
+
+    // Address + Read command
+    uint32_t read_byte_cmd = ((REVERT_24b_ADDR(addr & 0x00ffffff) << 8) | FC_RD);
+    // Load command to TX FIFO
+    spi_write_word(&spi, read_byte_cmd);
+    spi_wait_for_ready(&spi);
+
+    // Set up segment parameters -> send command and address
+    const uint32_t cmd_read_1 = spi_create_command((spi_command_t){
+        .len        = 3,                 // 4 Bytes
+        .csaat      = true,              // Command not finished
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirTxOnly      // Write only
+    });
+    // Load segment parameters to COMMAND register
+    spi_set_command(&spi, cmd_read_1);
+    spi_wait_for_ready(&spi);
+
+    // Set up segment parameters -> read length bytes
+    const uint32_t cmd_read_2 = spi_create_command((spi_command_t){
+        .len        = length-1,          // len bytes
+        .csaat      = false,             // End command
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirRxOnly      // Read only
+    });
+    spi_set_command(&spi, cmd_read_2);
+    spi_wait_for_ready(&spi);
+
+    /*
+     * Set RX watermark to length. The watermark is in words.
+     * If the length is not a multiple of 4, the RX watermark is set to length/4+1
+     * to take into account the extra bytes.
+     * If the length is higher then the RX FIFO depth, the RX watermark is set to
+     * RX FIFO depth. In this case the flag is not set to 0, so the loop will
+     * continue until all the data is read.
+    */
+    printf("Before data start arriving...\n");
```

Comment:
let's avoid having printf in here... for you it is OK to debug, you can just comment it, and leave it there

Reply:
Yes I imagined. I just wanted to debug the last things and then I'll remove all the printf from the BSP.